### PR TITLE
Support offset querying to enable implementation of pagination

### LINKED ISSUE #471: Add support for sorted, paginated and filtered index in RESTful view
At the moment, `query` of the service that backs a resource (https://oxidizing.github.io/sihl/sihl/Sihl/Web/Rest/module-type-SERVICE/index.html) doesn't support filtering, sorting and pagination. We want something like 

```ocaml
  val search
    :  ?sort:[ `Desc | `Asc ]
    -> ?filter:string
    -> int
    -> (t list * int) Lwt.t
```

that covers 80% of the needs of viewing and browsing a collection.

### PR DESCRIPTION
**Pull request by jerben**
Preparation for #471

### CODE REVIEW

File: sihl/src/contract_database.ml [hunk 1]
```diff
@@ -8,29 +8,81 @@ let name = "database"
...
-    -> string
-    -> string
-    -> string
+  (** ['a search_request] is a prepared SQL statement that can be used to sort,
+      filter and paginate (= search) a collection. *)
+  type 'a prepared_search_request
+
+  val prepare_requests : string -> string -> string -> string
+    (* Deprecated in 0.6.0 *)
+    [@@deprecated "Use prepare_search_request instead"]
```
Comment (jerben) [inline sihl/src/contract_database.ml:17]: @aronerben Let's do all deprecations like this. It won't compile when using it, but the compiler can be used to provide an upgrade path.

Comment (aronerben) [reply sihl/src/contract_database.ml:17]: @jerben Good approach!

File: sihl/src/index.mld [hunk 2]
```diff
@@ -731,6 +731,77 @@ If you have complex pre-conditions, you should move the service calls to the [da
...
+  let search sort filter ~limit ~offset =
+    Sihl.Database.query (fun connection ->
+        let module Connection = (val connection : Caqti_lwt.CONNECTION) in
+        Sihl.Database.run_search_request
+          connection
+          request
+          sort
+          filter
+          ~limit
+          ~offset)
```
Comment (aronerben) [inline sihl/src/index.mld:799]: Maybe mention somewhere that `offset` is slow?

**Review by aronerben — COMMENTED** [review]
Please check

File: sihl/src/contract_database.ml [hunk 3]
```diff
@@ -8,29 +8,81 @@ let name = "database"
 exception Exception of string
 
 module type Sig = sig
-  val prepare_requests
-    :  string
-    -> string
-    -> string
-    -> string
+  (** ['a search_request] is a prepared SQL statement that can be used to sort,
```
[the remark this answers is not in the document]
Comment (jerben) [reply sihl/src/contract_database.ml:11]: Fixed.

File: sihl/src/contract_database.ml [hunk 4]
```diff
@@ -8,29 +8,81 @@ let name = "database"
...
-       * ('c * 'a, 'b, [< `Many | `One | `Zero ]) Caqti_request.t
-       * ('c * 'a, 'b, [< `Many | `One | `Zero ]) Caqti_request.t
-       * (unit, int, [< `One ]) Caqti_request.t
+    -> 'a prepared_search_request
     -> [< `Asc | `Desc ]
     -> 'c option
     -> 'a
     -> ('b list * int) Lwt.t
+    (* Deprecated in 0.6.0 *)
+    [@@deprecated "Use prun_search_request instead"]
```
[the remark this answers is not in the document]
Comment (jerben) [reply sihl/src/contract_database.ml:56]: Fixed.

File: sihl/src/contract_database.ml [hunk 5]
```diff
@@ -8,29 +8,81 @@ let name = "database"
...
+      {!prepare_search_request}.
+
+      [sort] is the sort order. The field that is sorted by was set by
+      {!prepare_search_request}.
+
+      [filter] is an optional keyword that is used to filter the collection. If
+      no filter is provided, the collection is not filtered.
+
+      [offset] is the number of items that the returned partial result is
+      offseted by.
```
[the remark this answers is not in the document]
Comment (jerben) [reply sihl/src/contract_database.ml:73]: Fixed.

File: sihl/src/contract_database.ml [hunk 6]
```diff
@@ -8,29 +8,81 @@ let name = "database"
...
+
+      [limit] is the number of items of the returned partial result.
+
+      [offset] and [limit] can be used together to implementation pagination. *)
```
[the remark this answers is not in the document]
Comment (jerben) [reply sihl/src/contract_database.ml:77]: Fixed.

File: sihl/src/index.mld (hunk 2 above)
Comment (jerben) [reply sihl/src/index.mld:799]: Added a note.

Comment (aronerben) [reply sihl/src/index.mld:799]: Thanks!

**Pull request merged**